apps/astarte_appengine_api/lib/astarte_appengine_api/groups/queries.ex

```diff
@@ -541,4 +437,92 @@ defmodule Astarte.AppEngine.API.Groups.Queries do
         {:error, :database_error}
     end
   end
+
+  def check_all_devices_exist(realm_name, device_ids, group_changeset) do
+    keyspace = DataAccessRealm.keyspace_name(realm_name)
+
+    device_ids
+    |> Enum.chunk_every(100)
+    |> Enum.reduce_while(:ok, fn id_chunk, :ok ->
+      existing_ids =
+        from(d in DataBaseDevice,
+          prefix: ^keyspace,
+          where: d.device_id in ^id_chunk,
+          select: d.device_id
+        )
+        |> Repo.all()
+
+      if Enum.count(existing_ids) == Enum.count(id_chunk) do
+        {:cont, :ok}
+      else
+        # Some device_id was not present in the database. Take the first.
+        not_found =
+          id_chunk
+          |> Enum.find(&(&1 not in existing_ids))
+          |> Device.encode_device_id()
+
+        group_changeset =
+          group_changeset |> Changeset.add_error(:devices, "must exist (#{not_found} not found)")
+
+        {:halt, {:error, group_changeset}}
+      end
+    end)
+  end
+
+  def check_group_exists(realm_name, group_name) do
+    keyspace = DataAccessRealm.keyspace_name(realm_name)
+
+    from(GroupedDevice, select: [:group_name], limit: 1)
+    |> Repo.fetch_by([group_name: group_name], prefix: keyspace)
+  end
+
+  def add_to_grouped_device(realm_name, group_name, decoded_device_ids) do
+    keyspace = DataAccessRealm.keyspace_name(realm_name)
+    grouped_device_table = GroupedDevice.__schema__(:source)
+
+    insert_grouped_device_sql = """
+      INSERT INTO #{keyspace}.#{grouped_device_table} (group_name, insertion_uuid, device_id)
+      values (?, ?, ?)
+    """
+
+    queries =
+      decoded_device_ids
+      |> Enum.flat_map(fn device_id ->
+        insertion_uuid = UUID.uuid1()
+
+        group = %{group_name => insertion_uuid}
+
+        query =
+          from(DataBaseDevice, prefix: ^keyspace, where: [device_id: ^device_id])
+          |> update([d], set: [groups: fragment("groups + ?", ^group)])
+
+        update_device_groups =
+          Repo.to_sql(:update_all, query)
+
+        insert_grouped_device_params = [group_name, insertion_uuid, device_id]
+        insert_grouped_device = {insert_grouped_device_sql, insert_grouped_device_params}
```

Comment:
If we can merge https://github.com/astarte-platform/astarte/pull/1118 (@Annopaolo :wink:), this could be rewritten as 

```suggestion
        grouped_device =
          %GroupedDevice{
            group_name: group_name,
            insertion_uuid: insertion_uuid,
            device_id: device_id
          }
          
        insert_grouped_device = Repo.insert_to_sql(grouped_device, prefix: keyspace)
```

of course, deleting the `insert_grouped_device_sql` string from line 483